_episodes/02-keras.md

```diff
@@ -439,61 +493,163 @@ sns.lineplot(x=history.epoch, y=history.history['loss'])
+somewhat arbitrary and more careful consideration needs to be taken to
+pick hyperparameter values.
+
+## 8. Measuring Performance
+Now that we have a trained neural network it is important to assess how well it performs.
+We want to know how well it will perform in a realistic prediction scenario, measuring
+performance will also come back when tuning the hyperparameters.
+
+We have created a test set during the data preparation stage which we will use
+now to create a confusion matrix.
+
+### Predict the species of the test set
+The first step here is to use the neural network to predict the species of the test set
+using the `predict` function. This will return a `numpy` matrix, which I like to convert
+to a pandas dataframe to easily see the labels.
+~~~
+y_pred = model.predict(X_test)
+prediction = pd.DataFrame(y_pred, columns=target.columns)
+prediction
+~~~
+{:.language-python}
+
+Remember that the output of the network uses the `softmax` activation function and has three
+outputs, one for each species. This dataframe shows this nicely.
+
+We now need to transform this output to one penguin species per sample.
+We can do this by looking for the index of highest valued output and converting that
+to the corresponding species.
+Pandas dataframes have the `idxmax` function, which will do exactly that.
+
+~~~
+predicted_species = prediction.idxmax(axis="columns")
+predicted_species
+~~~
+{:.language-python}
+
+### Confusion matrix
+With the predicted species we can now create a confusion matrix and display it
+using seaborn.
+To create a confusion matrix we will use another convenient function from sklearn
+called `confusion_matrix`.
+This function takes as a first parameter the true labels of the test set.
+We can get these by using the `idxmax` method on the y_test dataframe.
+The second parameter is the predicted labels which we did above.
+
+~~~
+from sklearn.metrics import confusion_matrix
+
+true_species = y_test.idxmax(axis="columns")
+
+matrix = confusion_matrix(true_species, predicted_species)
+print(matrix)
+~~~
+{:.language-python}
+
+Unfortunately, this matrix is kinda hard to read. Its not clear which column and which row
+corresponds to which species.
+So let's convert it to a pandas dataframe with its index and columns set to the species
+as follows:
+
+~~~
+# Convert to a pandas dataframe
+confusion_df = pd.DataFrame(matrix, index=y_test.columns.values, columns=y_test.columns.values)
+
+# Set the names of the x and y axis, this helps with the readability of the heatmap.
+confusion_df.index.name = 'True Label'
+confusion_df.columns.name = 'Predicted Label'
+
+sns.heatmap(confusion_df, annot=True)
+~~~
+{:.language-python}
+
+We can then use the `heatmap` function from seaborn to create a nice visualization of
+the confusion matrix.
+The `annot=True` parameter here will put the numbers from the confusion matrix in
+the heatmap.
+
+~~~
+sns.heatmap(confusion_df, annot=True)
+~~~
+{:.language-python}
+
```

Comment:
I would show the output image of the heatmap here, instead of in the exercise (as you give the code already above the exercise). The exercise is then just about interpreting the results.